pkg/analyze/selector/engine/iterator.go

```diff
@@ -0,0 +1,147 @@
+// Package engine provides utilities for the parent selector package. It should
+// not be used directly outside of that package.
+//
+// The package proviedes extracts sets of elements from provided sets of named
+// sets such that all elements are in relation with one another for all
+// provided relations. Those relations are represented as functions from
+// pairs of elements from diffrent or same sets to a boolean value.
+package engine
+
+import (
+	"cmp"
+	"fmt"
+	"maps"
+	"slices"
+)
+
+// An Iterator provides means to iterate over all solutions.
+type Iterator struct {
+	labels    []string
+	relations Relations
+	values    map[string][]any
+}
+
+func NewIterator(relations Relations, values map[string][]any) (*Iterator, error) {
+	labels := relations.List()
+
+	for _, label := range labels {
+		if _, contains := values[label]; !contains {
+			return nil, fmt.Errorf("Missing key %s", label)
+		}
+	}
+
+	slices.SortFunc(labels, func(lhs, rhs string) int {
+		return cmp.Compare(len(values[lhs]), len(values[rhs]))
+	})
+
+	return &Iterator{
+		labels:    labels,
+		relations: relations,
+		values:    values,
+	}, nil
+}
+
+func (it *Iterator) ForEach(callback func(map[string]any) (bool, error)) error {
+	_, err := it.forEach(make(map[string]any, len(it.labels)), callback)
+	return err
+}
+
+func (it *Iterator) forEach(
+	prefix map[string]any,
+	callback func(map[string]any) (bool, error),
+) (bool, error) {
+	if len(prefix) >= len(it.labels) {
+		return callback(maps.Clone(prefix))
+	}
+
+	label := it.labels[len(prefix)]
+	for _, value := range it.values[label] {
+		prefix[label] = value
+
+		canAppend, err := it.canAppend(prefix, label, value)
+		if err != nil {
+			return false, err
+		}
+
+		if canAppend {
+			continu, err := it.forEach(prefix, callback)
```

Comment:
rename this variable. I can't think of a good name, maybe something along the lines of `shouldContinue`, `reachedEnd`, `proceed`?